clang/lib/Frontend/CompilerInvocation.cpp

```diff
@@ -3618,6 +3618,27 @@ void CompilerInvocationBase::GenerateLangArgs(const LangOptions &Opts,
     GenerateArg(Consumer, OPT_frandomize_layout_seed_EQ, Opts.RandstructSeed);
 }
 
+static void CheckBoundsSafetyLang(InputKind IK, DiagnosticsEngine &Diags) {
```

Comment:
I think it's better to handle this language check in Driver, similar to `OPT_fminimize_whitespace` but for all `Inputs`.

 It's not necessary to mention 

> This is especially important because some build systems, like xcbuild and somewhat clumsy Makefiles, will pass
C_FLAGS to Clang while building assembly files.

since this is general and applies to other options. No need to emphasize it specifically for bounds checking.

Reply:
> I think it's better to handle this language check in Driver, similar to OPT_fminimize_whitespace but for all Inputs.

@MaskRay If we do it in Driver, wouldn't the option be silently ignored for unsupported languages when we invoke `clang -cc1` directly?

Reply:
IMHO, that might be problematic because people would reasonably expect bounds-safety would work for C++/Obj-C/Obj-C++.

Reply:
There are various different options here.

1. The frontend silently accepts`-fexperimental-bounds-safety` with `-x c++` but doesn't actually enable `fexperimental-bounds-safety`.
2. The frontend accepts`-fexperimental-bounds-safety` with `-x c++` enables `-fexperimental-bounds-safety` and probably crashes because it's not supported.
3. The frontend rejects`-fexperimental-bounds-safety` with `-x c++` with an error.

I don't like (1.) because when writing frontend only tests it is inviting someone to accidentally write a test incorrectly because the frontend would never complain about the invalid configuration.

I don't like (2.) because crashing really isn't good when it could so easily be prevented.

I think we should do (3.) because it has none of the previously mentioned downsides.

Reply:
To be clear. I think it's fine to also do the language check in the driver but I think the frontend should **also** do language check so that it is not possible to run it in an invalid configuration.

Reply:
> I think it's better to handle this language check in Driver, similar to `OPT_fminimize_whitespace` but for all `Inputs`.

@MaskRay Thanks for the review! I made it a driver error here: 
[f125532](https://github.com/llvm/llvm-project/pull/70480/commits/f125532235d5120027ec261204ae09d315d0fa14)

We also keep the -cc1 check in order to prevent it from running with an unsupported mode (we still need some checks in cc1 anyway in case we want to silently ignore the option, instead of triggering the broken mode).